hazelcast/src/main/java/com/hazelcast/sql/impl/plan/cache/PlanCache.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.plan.cache;
+
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Cache for plans.
+ */
+public class PlanCache implements CachedPlanInvalidationCallback {
+
+    private final int maxSize;
+    private final ConcurrentHashMap<PlanCacheKey, CacheablePlan> plans = new ConcurrentHashMap<>();
+
+    public PlanCache(int maxSize) {
+        assert maxSize > 0;
+
+        this.maxSize = maxSize;
+    }
+
+    public CacheablePlan get(PlanCacheKey key) {
+        CacheablePlan plan = plans.get(key);
+
+        if (plan != null) {
+            plan.onPlanUsed();
+
+            return plan;
+        } else {
+            return null;
+        }
+    }
+
+    public void put(PlanCacheKey key, CacheablePlan plan) {
+        plans.put(key, plan);
+
+        plan.onPlanUsed();
+
+        shrinkIfNeeded();
+    }
+
+    public void invalidate(CacheablePlan plan) {
+        remove(plan);
+    }
+
+    public void clear() {
+        plans.clear();
+    }
+
+    public int size() {
+        return plans.size();
+    }
+
+    public void check(PlanCheckContext context) {
+        plans.values().removeIf(plan -> !plan.isPlanValid(context));
+    }
+
+    private void shrinkIfNeeded() {
+        int oversize = plans.size() - maxSize;
+
+        if (oversize <= 0) {
+            return;
+        }
+
+        // Sort plans according to their last used timestamps
+        TreeMap<Long, CacheablePlan> sorted = new TreeMap<>();
+
+        for (CacheablePlan plan : plans.values()) {
```

Comment:
We had similar use case in Jet. What we did is that when the maximum number was exceeded, we sorted the data and deleted 100 items at once.

Reply:
We assume that normally the number of plans is limited and normally should not exceed the maximum size. So the slow path is almost never invoked. If this happens, then a user is in trouble anyway, because planning takes significant time. For example, this may happen if a user uses literals with concatenation instead of parameters. 

This assumption means that there is a high hit rate. We would like to make plan lookup as cheap as possible. Therefore, `ConcurrentSkipListMap` is not an option: we use LRU algorithm and with `ConcurrentSkipListMap` we would have to update this data structure on every access.